src/vs/workbench/contrib/notebook/browser/viewModel/notebookOutlineEntryFactory.ts

```diff
@@ -47,23 +47,31 @@ export class NotebookOutlineEntryFactory {
 		let hasHeader = false;
 
 		if (isMarkdown) {
+			const cache = this.cachedMarkdownOutlineEntries.get(cell);
+			const markdownEntries: OutlineEntry[] = [];
 			const fullContent = cell.getText().substring(0, 10000);
-			for (const { depth, text } of getMarkdownHeadersInCell(fullContent)) {
+			const headers = cache?.alternativeId === cell.getAlternativeId() ? cache.headers : Array.from(getMarkdownHeadersInCell(fullContent));
```

Comment:
This caching results in saving 50% of the time used to calculate outlines.
Down from 50ms to 26

I.e. this PR brings down the total time in keystokes from 150ms to 0 (when debounced) and when user stops typing, then we spend 20ms.

We should see some significant perf gains here.